src/dispatching/dispatchers/filter.rs

```diff
@@ -0,0 +1,382 @@
+//! A dispatcher based on filters.
+
+use futures::StreamExt;
+
+use crate::{
+    dispatching::{filters::Filter, ErrorHandler, Handler, Updater},
+    types::{
+        CallbackQuery, ChosenInlineResult, InlineQuery, Message, Update,
+        UpdateKind,
+    },
+};
+
+type FilterWithHandler<'a, T, E> =
+    (Box<dyn Filter<T> + 'a>, Box<dyn Handler<T, E> + 'a>);
+type FiltersWithHandlers<'a, T, E> = Vec<FilterWithHandler<'a, T, E>>;
+
+/// A dispatcher based on filters.
+///
+/// Filters and handlers are executed in order of registering. The pseudocode
+/// looks like this:
+///
+/// ```no
+/// for pair in handlers_and_filters {
+///     if pair.filter.test(update) {
+///         pair.handle(update);
+///         return;
+///     }
+/// }
+///
+/// log("unhandeled update: " + update);
+/// ```
+///
+/// ## Examples
+///
+/// Simplest example:
+/// ```no_run
+/// # async fn run() {
+/// use std::convert::Infallible;
+///
+/// use teloxide::{
+///     dispatching::{
+///         dispatchers::filter::FilterDispatcher, updaters::polling_basic,
+///     },
+///     types::Message,
+///     Bot,
+/// };
+///
+/// async fn handle_edited_message(mes: Message) -> Result<(), Infallible> {
+///     println!("Edited message: {:?}", mes);
+///     Ok(())
+/// }
+///
+/// let bot = Bot::new("TOKEN");
+///
+/// // create dispatching which handlers can't fail
+/// // with error policy that just ignores all errors (that can't ever happen)
+/// let mut dp = FilterDispatcher::<Infallible, _>::new(|_| async {})
+///     // Add 'handler' that will handle all messages sent to the bot
+///     .message_handler(true, |mes: Message| async move {
+///         println!("New message: {:?}", mes);
+///         Ok(())
+///     })
+///     // Add 'handler' that will handle all
+///     // messages edited in chat with the bot
+///     .edited_message_handler(true, handle_edited_message);
+///
+/// // Start dispatching updates from long polling
+/// dp.dispatch(polling_basic(&bot)).await;
+/// # }
+/// ```
+///
+/// [`std::fmt::Debug`]: std::fmt::Debug
+/// [updater]: crate::dispatching::updater
+pub struct FilterDispatcher<'a, E, Eh> {
+    message_handlers: FiltersWithHandlers<'a, Message, E>,
+    edited_message_handlers: FiltersWithHandlers<'a, Message, E>,
+    channel_post_handlers: FiltersWithHandlers<'a, Message, E>,
+    edited_channel_post_handlers: FiltersWithHandlers<'a, Message, E>,
+    inline_query_handlers: FiltersWithHandlers<'a, InlineQuery, E>,
+    chosen_inline_result_handlers:
+        FiltersWithHandlers<'a, ChosenInlineResult, E>,
+    callback_query_handlers: FiltersWithHandlers<'a, CallbackQuery, E>,
+    error_handler: Eh,
+}
+
+/// An error produced either from [`Updater`] or [`Handler`].
+///
+/// [`Updater`]: crate::dispatching::Updater
+/// [`Handler`]: crate::dispatching::Handler
+pub enum ErrorKind<E1, E2> {
+    FromUpdater(E1),
+    FromHandler(E2),
+}
```

Comment:
Here I'm also not sure about naming. Maybe `DpError::{Updater, Handler}`? I think "kind" is a bit wrong cause it commonly just a part of something bigger. And "from" parts are just you know... redundant? And I would also rename generics: `E1 => UE`, `E2 => HE`

Reply:
But `DpError` is like "The error of a dispatcher", which is wrong. Any other names?